spec/bucket_store_integration_spec.rb

```diff
@@ -24,55 +26,129 @@
-      expect(described_class.for("#{base_bucket_uri}/prefix/file2").list.to_a.size).to eq(2)
-      expect(described_class.for("#{base_bucket_uri}/prefix/").list.to_a.size).to eq(203)
-
-      # List (without prefixes) should return everything
-      expect(described_class.for(base_bucket_uri.to_s).list.to_a).
-        to match_array(file_list.map { |filename| "#{base_bucket_uri}/prefix/#{filename}" })
-
-      # We know the content of the file, we can check `.download` returns it as expected
-      all_files = file_list.map do |filename|
-        [filename, "#{base_bucket_uri}/prefix/#{filename}"]
-      end
-      all_files.each do |content, key|
-        expect(described_class.for(key).download[:content]).to eq(content)
+      context "when handling streaming uploads / downloads" do
+        after do
+          file_to_read.close
+          file_to_write.close
+        end
+
+        shared_examples "consistently handles different file modes" do
+          it "has a consistent streaming interface" do
+            # Most common case for this API is streaming files to / from disk.
+            # So upload some files from local disk and also read into some as well
+            file_to_read.write(input_bytes)
+            file_to_read.rewind
+            filename = "streaming_test_file"
+
+            described_class.for("#{base_bucket_uri}/prefix/#{filename}").
+              stream.
+              upload!(file: file_to_read)
+
+            # Let's first try to grab it into a StringIO buffer and see if the
+            # the contents are equivalent
+            described_class.for("#{base_bucket_uri}/prefix/#{filename}").
+              stream.
+              download(file: in_memory_buffer)
+            in_memory_buffer.rewind
+            expect(in_memory_buffer.read).to eq(input_bytes)
+
+            # Now let's try writing to another temp file
+            described_class.for("#{base_bucket_uri}/prefix/#{filename}").
+              stream.
+              download(file: file_to_write)
+
+            file_to_write.rewind
+            expect(file_to_write.read).to eq(input_bytes)
+
+            # Santity check the list API behaves / interplays well with the streaming API
+            expect(described_class.for(base_bucket_uri.to_s).list.to_a.size).to eq(1)
+
+            # Delete all the files, the bucket should be empty afterwards
+            described_class.for(base_bucket_uri.to_s).list.each do |key|
+              described_class.for(key).delete!
+            end
+            expect(described_class.for(base_bucket_uri.to_s).list.to_a.size).to eq(0)
+          end
+        end
+
+        # Ruby files have a binary / textual mode.
+        # This means if you try and stick a byte array
+        # into a file it'll silently encode the bytes to
+        # whatever encoding is setup in the file.
+        # To validate byte arrays work we create binmode
+        # versions of input / output and the buffer.
```

Comment:
We can go a bit wider here, would be slightly easier to read

Reply:
Updated